Implementation of window function nth_value

ryzuo opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Implement window function `nth_value(expr, N)` that returns the value of the Nth row of the window frame.

Detailed description / Documentation draft:
nth_value(expr, N) returns the value of expr from the N-th row of the window frame. If there is no such row, the return value is NULL.


Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

Commit:
Implementaion of window function nth_value.

1. Implemented the window function nth_value
2. Make the return type of lag/lead, and nth_value functions to be
ColumnNullable, thus the returned value can be null if it is out
of frame.

Commit:
Make tests of nth_value pass

Commit:
Add missing test back in 01591_window_function

ryzuo commented:
Related issue
#26112

akuzm commented on src/Processors/Transforms/WindowTransform.cpp:
```diff
@@ -1492,7 +1494,7 @@ struct WindowFunctionLagLeadInFrame final : public WindowFunction
 
     DataTypePtr getReturnType() const override
-    { return argument_types[0]; }
+    { return std::make_shared<DataTypeNullable>(argument_types[0]); }
```
Are you also fixing the lagInFrame nullability problems in this PR? Better make two separate ones.

This change looks incorrect, because the return type of lagInFrame doesn't have to be nullable, e.g. when a not-null default value is specified. The problems in  https://github.com/ClickHouse/ClickHouse/issues/26115 are partly caused by that we apply nullability handling for aggregate functions to window functions as well, when we shouldn't. The fix needs some refactoring that would introduce deeper separation between aggregate and window functions. I started working on it but was distracted by other tasks. Probably will continue next week.

So in the meantime, could you please remove the fix for lagInFrame from this PR, and keep only the new function nth_value? That function looks good.

ryzuo replied:
Thanks for your comments, you are right, the force of nullability here is not appropriate and incomplete I accidently pushed it, I will clean it up, and I'm still looking into this problem. Speaking of the non-null default value, I found the result is not correct if the given default value is an arithmetic expression with type cast required, e.g. if the default value for the 3rd arg is like value*1.5, it looks like over flows. I tracked all the way up to interpreter, and also with null value, my guess there might be something in parser when it deduce the type of the 3rd arg? But I haven't looked into that part yet. I'm new in CH, still trying to understand lots of things.

Commit:
Cleanup the incomplete fix of nullability return value of lagInFrame/leadInFrame

akuzm commented:
> @akuzm Hi Alexander, there are still some check failures do not seem to relate to the code, Could you please help to take a look? And is there any more comments and suggestions?

clang-tidy complains about your changes:
```
2021-07-21 02:40:24 /build/obj-x86_64-linux-gnu/../src/Processors/Transforms/WindowTransform.cpp:1624:18: error: invalid case style for local variable 'srcColumnPtr' [readability-identifier-naming,-warnings-as-errors]
2021-07-21 02:40:24             auto srcColumnPtr = transform->blockAt(target_row).input_columns[workspace.argument_column_indices[0]];
2021-07-21 02:40:24                  ^~~~~~~~~~~~
2021-07-21 02:40:24                  src_column_ptr
```

Other things are probably not relevant, I'll check them before merging.

ryzuo commented:
I see, I will update that, thanks for your comments. And I saw your PR for 26115, should I change the nth_value returning in this PR like your changes? Or I should wait you merge both and open another one?

akuzm commented:
> I see, I will update that, thanks for your comments. And I saw your PR for 26115, should I change the nth_value returning in this PR like your changes? Or I should wait you merge both and open another one?

I just merged it. It would be good if you merge with master and change the nth_value implementation in this PR as well. By the way, I'm open for suggestions/test cases regarding the `lagInFrame` fix -- not sure I have it all totally figured out yet...

Commit:
Update the implementaion of nth_value

Make nth_value for nullable values for out of frame rows, as the same
fashion as lagInFrame just does.

akuzm commented on src/Processors/Transforms/WindowTransform.cpp:
```diff
@@ -1565,6 +1564,80 @@ struct WindowFunctionLagLeadInFrame final : public WindowFunction
+                "Function {} cannot be parameterized", name_);
+        }
+
+        if (argument_types.empty())
+        {
+            throw Exception(ErrorCodes::BAD_ARGUMENTS,
+                "Function {} takes at least one argument", name_);
+        }
+
```
If it always needs two arguments, we probably don't need this check, the next one is enough.

akuzm reviewed (approved):
LGTM other than that one comment.

akuzm commented on src/Processors/Transforms/WindowTransform.cpp:
```diff
@@ -9,7 +9,6 @@
 #include <Interpreters/ExpressionActions.h>
 #include <Interpreters/convertFieldToType.h>
 
-
```
We normally have two empty lines here, let's keep it.

Commit:
Remove unnecessary emtpy argument check for nth_value as code review suggested

akuzm commented:
The tests are flaky/broken in master.

alexey-milovidov commented:
Unfortunately, this function contains a major security issue: #29347

alexey-milovidov commented on tests/queries/0_stateless/01591_window_functions.sql:
```diff
@@ -403,6 +403,40 @@ window w as (order by number range between 1 preceding and 1 following)
 order by number
 ;
 
+-- nth_value without specific frame range given
```
New test should be created instead of adding new cases to old test.
